packages/apps/tools/src/components/Global/ModuleExplorer/index.tsx

```diff
@@ -1,89 +1,88 @@
-  onTabClose: IEditorProps['onTabClose'];
+  // onTabClose: IEditorProps['onTabClose'];
+  items: ISidePanelProps<IncompleteModuleModel>['data'];
+  onReload: ISidePanelProps<IncompleteModuleModel | Outline>['onReload'];
+  onExpandCollapse: ISidePanelProps<
+    IncompleteModuleModel | Outline
+  >['onExpandCollapse'];
 }
 
 const ModuleExplorer = ({
-  modules,
-  openedModules: fetchedModules,
-  onModuleClick,
-  onInterfaceClick,
-  onInterfacesExpand,
-  onModuleExpand,
+  items,
+  onReload,
+  onExpandCollapse,
   onActiveModuleChange,
-  onTabClose,
-}: IModuleExplorerProps): React.JSX.Element => {
+  openedModules: _openedModules,
+}: IModuleExplorerProps) => {
   const [activeModule, setActiveModule] = useState<IChainModule>();
-  const results = getModulesMap(modules);
   const [openedModules, setOpenedModules] =
-    useState<IChainModule[]>(fetchedModules);
+    useState<IChainModule[]>(_openedModules);
+  let outlineItems: TreeItem<Outline>[] = [];
 
-  const updateOpenedModules = useCallback(
-    (result: IChainModule) => {
-      setOpenedModules((prev) => {
-        const alreadyOpened = prev.find((openedModule) => {
-          return (
-            openedModule.moduleName === result.moduleName &&
-            openedModule.chainId === result.chainId &&
-            openedModule.network === result.network
-          );
-        });
-
-        if (alreadyOpened) {
-          return prev;
-        }
-
-        let add = result;
-
-        const enhanced = modules.find((module) => {
-          return (
-            module.moduleName === result.moduleName &&
-            module.chainId === result.chainId
-          );
-        });
-
-        if (enhanced) {
-          add = enhanced;
-        }
-
-        return [...prev, add];
-      });
-    },
-    [modules],
-  );
+  if (activeModule) {
+    outlineItems = chainModuleToOutlineTreeItems(activeModule, items);
+  }
 
   return (
     <div className={containerStyle}>
-      <SidePanel
-        results={results}
-        onResultClick={(result) => {
-          updateOpenedModules(result);
-          onModuleClick(result);
-          setActiveModule(result);
-        }}
-        onInterfaceClick={(result) => {
-          updateOpenedModules(result);
-          onInterfaceClick(result);
-          setActiveModule(result);
+      <SidePanel<IncompleteModuleModel | Outline>
+        data={[
+          {
+            children: items,
+            title: 'Explorer',
+            key: 'explorer',
+            label: 'Explorer',
+            data: 'explorer',
+          },
+          {
+            title: 'Outline',
+            key: 'outline',
+            label: 'Outline',
+            children: outlineItems,
+            data: 'outline',
+          },
+        ]}
+        onReload={onReload}
+        onModuleClick={({ data }) => {
+          if (isModuleLike(data)) {
+            const chainModule: IChainModule = moduleModelToChainModule(data);
+            setActiveModule(chainModule);
```

Comment:
Do we really need to do these steps before checking if it's already opened?

Reply:
Which steps do you mean exactly? 😁

Reply:
Basically call these two functions:
`            const chainModule: IChainModule = moduleModelToChainModule(data);
            setActiveModule(chainModule);`
            
            If I read it this way it seems a lot of calculation is being done before deciding that you might already have done it and it's unnecessary.

Reply:
Nothing unneccessary happens. When a module is being clicked, we always need to update the active module (`setActiveModule`). Then we add the clicked module to the `openedModules` (which is an array being used in the editor part, e.g. the tabs), but only if it wasn't already opened before. 

Does that make sense?

Reply:
Ah thinking about the other part you are going to build with the tabs it makes sense indeed. Thanks!